cupy/core/_routines_math.pyx

```diff
@@ -216,32 +372,127 @@ cdef ndarray scan(ndarray a, ndarray out=None):
         raise TypeError('Input array should be 1D array.')
 
     cdef Py_ssize_t block_size = 256
-
     if out is None:
-        out = ndarray(a.shape, dtype=a.dtype)
+        out = cupy.core._ndarray_init(a.shape, a.dtype)
     else:
         if a.size != out.size:
             raise ValueError('Provided out is the wrong size')
 
-    kern_scan = _inclusive_scan_kernel(a.dtype, block_size)
+    kern_scan = _inclusive_scan_kernel(a.dtype, block_size, op)
     kern_scan(grid=((a.size - 1) // (2 * block_size) + 1,),
               block=(block_size,),
               args=(a, out),
               shared_mem=a.itemsize * block_size * 2)
 
     if (a.size - 1) // (block_size * 2) > 0:
         blocked_sum = out[block_size * 2 - 1:None:block_size * 2]
-        scan(blocked_sum, blocked_sum)
-        kern_add = _add_scan_blocked_sum_kernel(out.dtype)
+        scan(blocked_sum, op, dtype, blocked_sum)
+        kern_add = _add_scan_blocked_sum_kernel(out.dtype, op)
         kern_add(grid=((a.size - 1) // (2 * block_size),),
                  block=(2 * block_size - 1,),
                  args=(out,))
     return out
 
 
+cdef ndarray _batch_scan_op(ndarray a, op, dtype=None, out=None):
```

Comment:
`out` looks required to be `ndarray` type (according to [this line](https://github.com/cupy/cupy/pull/2907/files#diff-aa843213093afe1f96359cbe1628c4cbR417)).
How about removing the default values?
```suggestion
cdef ndarray _batch_scan_op(ndarray a, op, dtype, ndarray out):
```